File: Arrays/Algorithms/Searching/Sorted/meta_binary_search.py
import math

def meta_binary_search(arr, x):
	n = len(arr)
	lg = int(math.log2(n)) + 1
	pos = -1
	for i in range(lg, -1, -1):
		new_pos = pos + (1 << i)
		if new_pos < n and arr[new_pos] <= x:
			pos = new_pos
	if pos != -1 and arr[pos] == x:
		return pos
	return -1

if __name__ == "__main__":
	n, x = map(int, input().split())
	arr = list(map(int, input().split()))
	print(meta_binary_search(arr, x))

File: Arrays/Algorithms/Searching/Sorted/meta_binary_search.cpp
#include <bits/stdc++.h>
using namespace std;

int metaBinarySearch(const vector<int>& arr, int x) {
	int n = arr.size();
	int left = 0, right = n - 1;
	int lg = log2(n) + 1;
	int pos = -1;
	for (int i = lg; i >= 0; i--) {
		int newPos = pos + (1 << i);
		if (newPos < n && arr[newPos] <= x) pos = newPos;
	}
	if (pos != -1 && arr[pos] == x) return pos;
	return -1;
}

int main() {
	int n, x;
	cin >> n >> x;
	vector<int> arr(n);
	for (int i = 0; i < n; i++) cin >> arr[i];
	cout << metaBinarySearch(arr, x) << endl;
	return 0;
}
